### PR TITLE
Detect missing dependencies between tasks

### LINKED ISSUE #15308: Deprecate missing dependencies between work items
When a work item (consumer) has the output of another work item (producer) as an input, then there also needs to be a dependency in the execution graph.

We should detect such a missing dependency and issue a deprecation warning.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by wolfs**
Detect missing dependencies between tasks by matching up their inputs and outputs.

Fixes #15308.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 1]
```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
...
+            return false;
+        }
+        if (consumer.getDependencySuccessors().contains(producer)) {
+            return false;
+        }
+        for (Node dependency : consumer.getAllSuccessors()) {
+            if (dependency == producer || dependency.getDependencySuccessors().contains(producer)) {
+                return false;
+            }
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:115]: What's the difference between searching in `getDependencySuccessors()` vs `getAllSuccessors()`? It seems we should not be doing both.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 2]
```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
...
+        // Do a deep search
+        ArrayDeque<Node> queue = new ArrayDeque<>();
+        consumer.getAllSuccessors().forEach(queue::add);
+        while (!queue.isEmpty()) {
+            Node dependency = queue.removeFirst();
+            if (dependency == producer) {
+                return false;
+            }
+            dependency.getAllSuccessors().forEach(queue::add);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:124]: Why not use `queue.addAll()`? Seems to be more optimized.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 3]
```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
...
+        }
+        return true;
+    }
+
+    private void emitMissingDependencyDeprecationWarning(Node producer, Node consumer) {
+        DeprecationLogger.deprecateBehaviour(String.format("%s consumes the output of %s, but does not declare a dependency.", consumer, producer)).willBeRemovedInGradle7().undocumented().nagUser();
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:130]: I think these calls are a lot more readable if you put each `.xxx()` call on a new line.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    private interface RelatedLocation {
+        Collection<Node> getNodes();
+        void visitNodes(VfsRelativePath relativePath, Consumer<Node> nodeConsumer);
+        void visitNodes(Consumer<Node> nodeConsumer);
+        RelatedLocation recordRelatedToNode(VfsRelativePath relativePath, Node node);
+
+        ChildMap<RelatedLocation> getChildren();
+    }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java:62]: I don't think we should have a private interface for a single private implementation.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+public class RelatedLocations {
+    private volatile RelatedLocation root = new DefaultRelatedLocation(EmptyChildMap.getInstance(), ImmutableList.of());
+
+    public Collection<Node> getNodesRelatedTo(String location) {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java:33]: Any reason not to return a `Set` or even an `ImmutableSet` here?

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import org.gradle.internal.snapshot.ChildMap;
+import org.gradle.internal.snapshot.ChildMapFactory;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+public class RelatedLocations {
+    private volatile RelatedLocation root = new DefaultRelatedLocation(EmptyChildMap.getInstance(), ImmutableList.of());
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java:31]: I wonder if it's better like this, or if we should be using `DefaultRelatedLocation` as the thing. For the VFS I see a reason for keeping a mutable wrapper around the immutable data structure, because we need to access the mutable thing from different places. Here we only have one place where we update this thing, and I think we could simplify significantly if we dropped the mutable wrapper.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+
+    public synchronized void clear() {
+        root = new DefaultRelatedLocation(EmptyChildMap.getInstance(), ImmutableList.of());
+    }
+
+    private interface RelatedLocation {
+        Collection<Node> getNodes();
+        void visitNodes(VfsRelativePath relativePath, Consumer<Node> nodeConsumer);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java:57]: The parameter `relativePath` should be named after its role/function.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        void visitNodes(Consumer<Node> nodeConsumer);
+        RelatedLocation recordRelatedToNode(VfsRelativePath relativePath, Node node);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java:59]: Same here.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:124]: Because `getAllSuccessors` is an `Iterable` and not a collection.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalBuildIntegrationTest.groovy [hunk 9]
```diff
@@ -995,6 +995,7 @@ task generate(type: TransformerTask) {
 """
 
         when:
+        executer.expectDeprecationWarning(":transform1 consumes the output of :src2, but does not declare a dependency. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0.")
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalBuildIntegrationTest.groovy]: I don't think the deprecation should be a failure in 7.0. Though the warning message will have 8.0 in it very soon.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java [hunk 10]
```diff
@@ -433,6 +429,7 @@ public void validate(WorkValidationContext validationContext) {
                 reservedFileSystemLocationRegistry,
                 typeValidationContext
             ));
+            context.getValidationAction().run();
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: I am passing the validation warnings now to the validation context.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 11]
```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
...
     }
+
+    private void detectMissingDependencies(LocalTaskNode node) {
+        RelatedLocations consumedDirectories = consumedAndProducedLocations.getConsumedDirectories();
+        for (String outputPath : node.getMutationInfo().outputPaths) {
+            consumedDirectories.getNodesRelatedTo(outputPath).stream()
+                .filter(consumerNode -> missesDependency(node, consumerNode))
+                .forEach(consumerWithoutDependency -> emitMissingDependencyDeprecationWarning(node, consumerWithoutDependency));
+        }
+        Set<String> consumedLocations = new LinkedHashSet<>();
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:73]: So we report the errors in the same order. I suppose we can use a regular `HashSet`, now that we use validation context, since the messages will be sorted there?

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:115]: `getAllSuccessors` includes `mustRunAfter` and finalizers as well. `getDependencySuccessors()` only includes actual hard dependencies.

The reason why we use both methods here is a performance optimization: We can easily check if something is part of the `dependencySuccessors`, since it is a set. `allSuccessors` is only an `Iterable`, so we need to iterate over it to check whether a node is in there or not.

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Comment by wolfs** [comment]
@bot-gradle test QFL

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 12]
```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
...
+                public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    consumedLocations.add(file.getAbsolutePath());
+                }
+            }));
+        consumedDirectories.recordRelatedToNode(node, consumedLocations);
+        for (String consumedLocation : consumedLocations) {
+            consumedAndProducedLocations.getProducedDirectories().getNodesRelatedTo(consumedLocation).stream()
+                .filter(producerNode -> missesDependency(producerNode, node))
+                .forEach(producerWithoutDependency -> emitMissingDependencyDeprecationWarning(producerWithoutDependency, node));
+        }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:101]: I am currently collecting the elements first and then iterate over them afterwards, so I only check each location once. Checking each location directly when I collect it caused a performance regression in `ArchiveTreePerformanceTest`, since we add the same location 1000 times as an input.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java]: The deprecation logger call is gone now.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java]: I removed the interface.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java (hunk 5 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java]: Returning an `ImmutableSet` now.

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Comment by wolfs** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/execution/plan/ConsumedAndProducedLocations.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+public class ConsumedAndProducedLocations {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/ConsumedAndProducedLocations.java:21]: I only use this type so I can inject the dependencies. I can't have two injectable things of the same type. The type is now only known in the service registries, and the two classes using it receive the two fields separately.

File: subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java (hunk 6 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java]: We still have multiple places where we update the thing: in the execution plan and from the task executor. So I think we need a mutable wrapper to keep the right reference in two places.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalBuildIntegrationTest.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalBuildIntegrationTest.groovy]: Is there a reason not to change it to `willBeRemovedInGradle8()` instead of `willBeRemovedInGradle7()` already?

[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalBuildIntegrationTest.groovy]: Yes, because it now runs through our validation infrastructure, so I can't change the `willBeRemovedInGradle7()` easily, since I would need to add a new flag for task validations.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:116]: Please add a comment about why we need this.

**Comment by ldaley** [comment]
What happens when we detect?

**Comment by lptr** [comment]
Deprecation warning, break in 8.0.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java]: Added.

**Comment by wolfs** [comment]
@bot-gradle test this

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy [hunk 14]
```diff
@@ -756,10 +755,12 @@ class RepositoriesDeclaredInSettingsIntegrationTest extends AbstractModuleDepend
...
-        failure.assertThatCause(containsNormalizedString("""Could not resolve all dependencies for configuration ':conf'.
-The project declares repositories, effectively ignoring the repositories you have declared in the settings.
-You can figure out how project repositories are declared by configuring your build to fail on project repositories.
-See https://docs.gradle.org/${GradleVersion.current().version}/userguide/declaring_repositories.html#sub:fail_build_on_project_repositories for details."""))
+        // FIXME wolfs
+//        failure.assertThatCause(containsNormalizedString("""Could not resolve all dependencies for configuration ':conf'.
+//The project declares repositories, effectively ignoring the repositories you have declared in the settings.
+//You can figure out how project repositories are declared by configuring your build to fail on project repositories.
+//See https://docs.gradle.org/${GradleVersion.current().version}/userguide/declaring_repositories.html#sub:fail_build_on_project_repositories for details."""))
+        failure.assertThatCause(containsNormalizedString("""Could not resolve all files for configuration ':conf'."""))
```
Comment (lptr) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy:763]: Why is this necessary?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+                }
+            }
+        """
+
+        expect:
+        succeeds("producer", "consumer")
+        succeeds("consumer", "producer")
+    }
+
+    def "dependency chain counts as dependency (including #dependency)"() {
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:94]: How about `transitive dependencies are accepted as valid dependencies`?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy (hunk 14 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy:763]: The error message for resolving the configuration as a file collection and calling some special methods on the configuration is different. @melix is looking into it.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+            d.dependsOn(c)
+            ${dependency}
+            b.dependsOn(a)
+
+            a.dependsOn(producer)
+        """
+
+        expect:
+        succeeds("producer", "b", "c", "consumer")
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:129]: Why don't we just invoke `consumer`?

File: subprojects/core/src/main/java/org/gradle/execution/plan/Node.java [hunk 17]
```diff
@@ -247,6 +247,11 @@ public void dependenciesProcessed() {
         return dependencySuccessors;
     }
 
+    @OverridingMethodsMustInvokeSuper
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/Node.java:250]: We are defining a new term now. Let's add some Javadoc describing its relationship to the other terms.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy (hunk 16 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:129]: Because for `finalizedBy` and `mustRunAfter`, the intermediate tasks are not triggered.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy (hunk 15 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy]: Renamed.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy (hunk 16 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:130]: I added a comment.

File: subprojects/core/src/main/java/org/gradle/execution/plan/Node.java (hunk 17 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/Node.java:255]: I added some Javadoc. Note that the term is not new, we already have `hasHardSuccessor()`.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy (hunk 16 above)
Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy:130]: Thanks. Another way to disambiguate would be if the list of tasks to execute would be a parameter to the test.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by ldaley** [comment]
ping on https://github.com/gradle/gradle/pull/15432#issuecomment-741511639

What is the point of the detection? What happens?

**Comment by wolfs** [comment]
> What is the point of the detection? What happens?

@ldaley Lorant answered above: We emit a deprecation warning and disable optimizations for the task for which we detect the missing dependency (like for all task validation problems). We plan on breaking the build in Gradle 8.0 for a missing dependency between tasks. Above all we want to point out this problem to the user so they can fix it.

**Comment by ldaley** [comment]
Sorry, I missed that comment.

**Comment by lptr** [comment]
The typical use case for the warning is when you have a producer and a consumer task connected by manually pointing them to the same location, but not adding a dependency. In such a case consumer can be executed at the same time or even before producer, which can cause corrupted outcome.
